[@mantine/form] use-form: Make functions referentially stable

mattrunyon opened the pull request:
Fixes #1196, #1193 

This makes all functions returned by `useForm` stable. I used the eslint react-hooks plugin for their exhaustive dependencies rule while converting these. It helped catch everywhere that the state needed to be switched to using the ref value within the `useCallback` uses

rtivital commented on src/mantine-form/src/use-form.ts:
```diff
@@ -63,46 +63,73 @@ export interface UseFormReturnType<T> {
+ * @param value The value to clone in the ref
+ * @returns The ref which clones the value
+ */
+function useStateRef<T>(value: T): React.MutableRefObject<T> {
```
It should be in a separate file

rtivital commented on src/mantine-form/src/use-form.ts:
```diff
@@ -63,46 +63,73 @@ export interface UseFormReturnType<T> {
+ * Creates a ref that stays up to date with the state of the param
+ * Used to help with all the useCallback signatures in useForm
+ *
+ * @param value The value to clone in the ref
```
No need for documentation in comments, it's pretty clear what hook does

rtivital commented on src/mantine-form/src/use-form.ts:
```diff
@@ -63,46 +63,73 @@ export interface UseFormReturnType<T> {
   schema,
 }: UseFormInput<T>): UseFormReturnType<T> {
   const [errors, setErrors] = useState(filterErrors(initialErrors));
+  const errorsRef = useStateRef(errors);
```
I think it is too wordy to create separate variables and calling hooks to get refs. Instead move this logic to `useRefState` hook, that will handle that, for example:

```tsx
const [errors, setErrors, errorsRef] = useRefState(filterErrors(initialErrors));
```

rtivital reviewed (commented):
I haven't checked how it works, but if tests are passing, then it is 👍

mattrunyon commented:
Made the changes you suggested (and also made another helper hook to track the state of a prop in a ref).

I also added a unit test for checking the functions are stable. Tested it failed when I removed `useCallback` from one of functions returned by the hook

rtivital commented:
Thanks!

rtivital commented:
I've merged it without checking how it works in real demos (stupid of me) and it turns out that it does not work at all – all demos in broken (values updates are random). I've reverted the changes, so if you want, you can open another PR for this.
